File: Pie02/Pie02IronPython/Expressions/MethodDeclaration.py
from Expression import Expression

class MethodDeclaration(Expression):
    """description of class"""

    def __init__(self, parentExpression):
        super(self.__class__, self).__init__(parentExpression)
        self.Modifiers = []
        self.ReturnType = None
        self.GenericList = None

    @staticmethod
    def Build(parser, parent, node):
        m = MethodDeclaration(parent)
        parent.Children.append(m)

        for modifier in node.ChildNodes[0].ChildNodes:
            m.Modifiers.append(modifier.FindTokenAndGetText())

        if node.ChildNodes[1].FindTokenAndGetText() != "void":
            m.ReturnType = Expression(m)
            parser.ConsumeParseTree(m.ReturnType, node.ChildNodes[1])

        m.Name = node.ChildNodes[2].FindTokenAndGetText()

        if node.ChildNodes[3].ChildNodes.Count > 0:
            m.GenericList = Expression(m)
            parser.ConsumeParseTree(m.GenericList, node.ChildNodes[3])

        parser.ConsumeParseTree(m, node.ChildNodes[7]);


    def Validate(self):
        # Validate modifiers
        self.Modifiers = [modifier.replace('final', 'sealed') for modifier in self.Modifiers]
        self.Modifiers = [modifier.replace('shared', 'static') for modifier in self.Modifiers]
        if self.Modifiers.count('internal') == 0 and self.Modifiers.count('public') == 0 and self.Modifiers.count('private') == 0: self.Modifiers.append('public')  
        super(self.__class__, self).Validate()    


    def Emit(self, source):
        for m in self.Modifiers:
            source += m
            source += " "

        if self.ReturnType == None:
            source += " void "
        else:
            source = self.ReturnType.Emit(source)

        source += self.Name
        if self.GenericList != None:
            source = self.GenericList.Emit(source)

        source += "("

        source += ")"

        source += "\n{\n"

        source = super(self.__class__, self).Emit(source)

        source += "}\n"
        return source

File: Pie02/Pie02CS/Expressions/MethodDeclaration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Irony.Parsing;

namespace Pie02CS.Expressions
{
    class MethodDeclaration
        : Expression
    {
        List<string> Modifiers = new List<string>();
        string MethodType = "";
        Expression Parameters;
        public bool Constructor = false;

        public MethodDeclaration(Expression parentExpression)
            : base(parentExpression)
        {
            Parameters = new Expression(this);
        }

        public static void Build(IronyParser parser, Expression parentExpression,ParseTreeNode node)
        {
            var m = new MethodDeclaration(parentExpression);
            parentExpression.Children.Add(m);

            foreach (var modifier in node.ChildNodes[0].ChildNodes)
                m.Modifiers.Add(modifier.FindTokenAndGetText());

            m.MethodType = node.ChildNodes[1].Term.ToString();

            if (m.MethodType != "new")
            {
                m.Name = node.ChildNodes[2].FindTokenAndGetText();

                parser.ConsumeParseTree(m.Parameters, node.ChildNodes[3]);

                parser.ConsumeParseTree(m, node.ChildNodes[4]);
            }
            else
            {
                m.Name = parentExpression.Name;
                parser.ConsumeParseTree(m.Parameters, node.ChildNodes[2]);

                parser.ConsumeParseTree(m, node.ChildNodes[3]);
            }


        }

        public override void Validate()
        {
            ScopeStack.Push(Scope);

            for (int i = 0; i < Modifiers.Count; i++)
            {
                switch (Modifiers[i])
                {
                    case "final":
                        Modifiers[i] = "sealed";
                        break;
                    case "shared":
                        Modifiers[i] = "static";
                        break;
                }
            }

            if(ParentExpression is ClassDeclaration)
            {
                var c = ParentExpression as ClassDeclaration;
                if (c.ClassType == "module" && !Modifiers.Contains("static"))
                    Modifiers.Add("static");
            }

            if (!Modifiers.Contains("internal") && !Modifiers.Contains("public") && !Modifiers.Contains("private")) Modifiers.Add("public");

            foreach(var c in Parameters.Children)
            {
                var p = c as Parameter;
                var v = new Variable();
                v.FullName = p.Identifier;
                //PropagateVariable(v);
                Scope.Variables.Add(v);
            }
            base.Validate();

            ScopeStack.Pop(Scope);
        }

        public override void Emit(ref string source)
        {
            foreach(var m in Modifiers)
            {
                source += m + " ";
            }

            if(MethodType != "new")
                if (MethodType == "act")
                    source += " void ";
                else
                    source += " dynamic ";

            source += Name;

            source += "(";

            if(Parameters.Children.Count > 0)
            {
                Parameters.Children[0].Emit(ref source);
                for(int i= 1; i < Parameters.Children.Count; i++)
                {
                    source += ", ";
                    Parameters.Children[i].Emit(ref source);
                }
            }

            source += ")";

            if(Children.Count > 0)
            {
                if(Children[0].Children[0] is MethodInvocation)
                {
                    if(Children[0].Children[0].Name == "base")
                    {
                        source += " : ";
                        Children[0].Children[0].Emit(ref source);
                    }
                }
            }

            source += "\n{\n";

            foreach(var c in Children)
            {
                if(c is Statement)
                {
                    var m = (c as Statement).Children[0];
                    if(m is MethodInvocation)
                    {
                        if (m.Name == "base") continue;
                    }
                }
               // else
              //  {
                    c.Emit(ref source);
                    source += ";\n";
               // }
            }

            source += "}\n";

        }
    }
}
